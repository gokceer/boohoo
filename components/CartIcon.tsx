import React from 'react'
import { TouchableOpacity, Text, StyleSheet } from 'react-native'
import { useNavigation } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack'
import { ShoppingBagIcon } from "react-native-heroicons/solid";
import { useAppSelector } from '../hooks'
import { selectCartItems } from "../slices/shoppingCartSlice"

import type { RootStackParamList } from "../App"

type ShoppingCartNavigationProp = NativeStackNavigationProp<RootStackParamList, 'ShoppingCart'>;

// This component shows cart icon, quantity and navigates to ShoppingCartScreen
const CartIcon = () => {
  const navigation = useNavigation<ShoppingCartNavigationProp>()
  const quantity = useAppSelector(selectCartItems);

  return (
    <TouchableOpacity testID='cart_btn' style={styles.quantityText} onPress={() => navigation.navigate("ShoppingCart")}>
      <ShoppingBagIcon testID='cart_icon' size={20} color="#841584" />
      <Text>{quantity.length}</Text>
    </TouchableOpacity>
  )
}

const styles = StyleSheet.create({
  quantityText: { flexDirection: "row", alignItems: "center" }
})

export default CartIcon