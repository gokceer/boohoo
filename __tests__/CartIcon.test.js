import { Provider } from "react-redux";
import { store } from "../store";
import { fireEvent, render } from "@testing-library/react-native";
import CartIcon from "../components/CartIcon";

const mockedNavigate = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

const setup = (props = {}) => {
  const wrapper = render(
    <Provider store={store}>
      <CartIcon {...props} />
    </Provider>
  );
  return wrapper;
};

describe("Cart icon unit tests", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  it("renders correctly, snapshot test", () => {
    const tree = setup();
    expect(tree).toMatchSnapshot();
  });

  it("Icon existance and quantity text test", () => {
    const wrapper = setup();
    expect(wrapper.getByTestId("cart_icon")).toBeTruthy();
    expect(wrapper.getByText("0")).toBeTruthy();
  });

  it("Touchable opacity onpress navigate to ShoppingCart screen test", () => {
    const wrapper = setup();
    const cartButton = wrapper.getByTestId("cart_btn");

    fireEvent.press(cartButton);

    expect(mockedNavigate).toHaveBeenCalledWith("ShoppingCart");
  });
});
