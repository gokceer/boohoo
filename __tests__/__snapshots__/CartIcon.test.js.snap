// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart icon unit tests renders correctly, snapshot test 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "opacity": 1,
    }
  }
  testID="cart_btn"
>
  <RNSVGSvgView
    align="xMidYMid"
    aria-hidden="true"
    bbHeight="20"
    bbWidth="20"
    color="#841584"
    fill="currentColor"
    focusable={false}
    height={20}
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 20,
          "width": 20,
        },
      ]
    }
    testID="cart_icon"
    tintColor="#841584"
    vbHeight={24}
    vbWidth={24}
    width={20}
  >
    <RNSVGGroup
      fill={
        Object {
          "type": 2,
        }
      }
      propList={
        Array [
          "fill",
        ]
      }
    >
      <RNSVGPath
        clipRule={0}
        d="M7.5 6v.75H5.513c-.96 0-1.764.724-1.865 1.679l-1.263 12A1.875 1.875 0 004.25 22.5h15.5a1.875 1.875 0 001.865-2.071l-1.263-12a1.875 1.875 0 00-1.865-1.679H16.5V6a4.5 4.5 0 10-9 0zM12 3a3 3 0 00-3 3v.75h6V6a3 3 0 00-3-3zm-3 8.25a3 3 0 106 0v-.75a.75.75 0 011.5 0v.75a4.5 4.5 0 11-9 0v-.75a.75.75 0 011.5 0v.75z"
        fillRule={0}
        propList={
          Array [
            "fillRule",
          ]
        }
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text>
    0
  </Text>
</View>
`;
